src/mem/epoch/tagged_atomic.rs

```diff
@@ -0,0 +1,301 @@
+use std::marker::PhantomData;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::{self, Ordering};
+
+use super::{Owned, Shared, Guard};
+
+/// Provides atomic access to a `(*const T, usize)` pair.
+///
+/// The pair consists of:
+///
+/// - a (nullable) pointer of type `T`, interfacing with the `Owned` and `Shared` types
+///
+/// - a `usize` tag value that can fit into the unused bits of the pointer:
+/// `tag < std::mem::align_of::<T>()`
+#[derive(Debug)]
+pub struct TaggedAtomic<T> {
+    ptr: atomic::AtomicUsize,
+    _marker: PhantomData<*const T>,
+}
+
+unsafe impl<T: Sync> Send for TaggedAtomic<T> {}
+unsafe impl<T: Sync> Sync for TaggedAtomic<T> {}
+
+/// Returns the power of two that the tag must be strictly
+/// less than. This value is `2.pow(n_unused_bits)`.
+fn tag_ceil<T>() -> usize {
+    0b1 << mem::align_of::<T>().trailing_zeros()
```

Comment:
Nit: If we keep this support for non-power-of-two alignments, I think it would be slightly cleaner to drop the `0b` prefix. Also, I think the documentation would be slightly clearer as something like:

Returns the largest power of two that divides the alignment of `T`. Any tag placed on a pointer to `T` must be strictly less than this value, so as to not interfere with any bits that the pointer actually uses.